### PR TITLE
[Merged by Bors] - Add a method for mapping `Mut<T>` -> `Mut<U>`

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

When designing an API, you may wish to provide access only to a specific field of a component or resource. The current options for doing this in safe code are

* `*Mut::into_inner`, which flags a change no matter what.
* `*Mut::bypass_change_detection`, which misses all changes.

## Solution

Add the method `map_unchanged`.

### Example

```rust
// When run, zeroes the translation of every entity.
fn reset_all(mut transforms: Query<&mut Transform>) {
    for transform in &mut transforms {
        // We pinky promise not to modify `t` within the closure.
        let translation = transform.map_unchanged(|t| &mut t.translation);
        // Only reset the translation if it isn't already zero.
        translation.set_if_not_equal(Vec2::ZERO);
    }
}
```

---

## Changelog

+ Added the method `map_unchanged` to types `Mut<T>`, `ResMut<T>`, and `NonSendMut<T>`.

### CODE REVIEW

**Review by afonsolage — APPROVED** [review]

**Comment by inodentry** [comment]
I don't quite understand why this is necessary? What are you planning to use it for?

**Comment by maniwani** [comment]
I think #4413 is related.

File: crates/bevy_ecs/src/change_detection.rs [hunk 1]
```diff
@@ -173,6 +173,17 @@ macro_rules! impl_into_inner {
                 self.set_changed();
                 self.value
             }
+
+            /// Maps to an inner value by applying a function to the contained reference, without flagging a change.
+            ///
+            /// You must *not* modify the argument passed to the closure.
+            /// Violating this rule will likely result in logic errors, but it will not cause undefined behavior.
+            pub fn map_unchanged<U: ?Sized>(self, f: impl FnOnce(&mut $target) -> &mut U) -> Mut<'a, U> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/change_detection.rs:181]: This sort of functional code can be hard to follow; can you add a doc test to demonstrate its usage and properties?

**Comment by alice-i-cecile** [comment]
Clever. If we can improve the docs I'm on board.

File: crates/bevy_ecs/src/change_detection.rs (hunk 1 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/change_detection.rs:181]: What do you think of the example I used in the PR desc?

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/change_detection.rs:181]: Quite good!

**Review by TheRawMeatball — APPROVED** [review]

File: crates/bevy_ecs/src/change_detection.rs [hunk 2]
```diff
@@ -173,6 +173,17 @@ macro_rules! impl_into_inner {
                 self.set_changed();
                 self.value
             }
+
+            /// Maps to an inner value by applying a function to the contained reference, without flagging a change.
+            ///
+            /// You must *not* modify the argument passed to the closure.
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/change_detection.rs:179]: Nit: The word "must" sounds a bit strong for something that doesn't cause UB. Maybe replace it with "should"?
```suggestion
            /// You should *not* modify the argument passed to the closure.
```
Or keep it if you feel it needs to be made a strong warning.

File: crates/bevy_ecs/src/change_detection.rs [hunk 3]
```diff
@@ -497,4 +508,35 @@ mod tests {
...
+        };
+        assert!(!ptr.is_changed());
+
+        // Perform a mapping operation.
+        let mut inner = ptr.map_unchanged(|x| &mut x.0);
+        assert!(!inner.is_changed());
+
+        // Mutate the inner value.
+        *inner = 64;
+        assert!(inner.is_changed());
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/change_detection.rs:540]: Just a question as I’m not super familiar with the change detection internals. Would this cause the outer container to also indicate a change?

I think it would be good to add another assertion to `outer` here to ensure it is/isn't changed.

**Review by MrGVSV — APPROVED** [review]

File: crates/bevy_ecs/src/change_detection.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/change_detection.rs:540]: It shouldn't. I agree with the request for that assert though.

**Comment by maniwani** [comment]
Should this method be `unsafe`? There's no memory safety issue, but the "caller pinky promises not to mutate" part seems like it shares the sentiment.

**Comment by JoJoJet** [comment]
> Should this method be `unsafe`? 

I don't think there's any need for it to be unsafe, but I'm willing to change it if the consensus disagrees with me.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by mockersf** [comment]
This is not `unsafe`, it's "if you do dumb things, dumb things will happen". Until we can declare our own function colours in Rust, we can't really mark those 🤷

**Review by mockersf — APPROVED** [review]

**Comment by mockersf** [comment]
@TheRawMeatball should this close #4413 if it's merged?
